client/src/test/java/io/pravega/client/state/impl/SynchronizerTest.java

```diff
@@ -411,5 +411,121 @@ public void testSetOperations() throws EndOfSegmentException {
+                new JavaSerializer<>(),
+                SynchronizerConfig.builder().build());
+
+        StateSynchronizer<RevisionedImpl> syncB = clientFactoryB.createStateSynchronizer(streamName,
+                new JavaSerializer<>(),
+                new JavaSerializer<>(),
+                SynchronizerConfig.builder().build());
+
+        assertEquals(0, syncA.bytesWrittenSinceCompaction());
+        assertEquals(0, syncB.bytesWrittenSinceCompaction());
+
+        AtomicInteger callCount = new AtomicInteger(0);
+
+        syncA.initialize(new RegularUpdate("a"));
+        syncB.initialize(new RegularUpdate("b"));
+        assertEquals("a", syncA.getState().getValue());
+        assertEquals("a", syncB.getState().getValue());
+
+        syncA.updateState((state, updates) -> {
+            callCount.incrementAndGet();
+            updates.add(new RegularUpdate("b"));
+        });
+        assertEquals(1, callCount.get());
+        assertEquals("b", syncA.getState().getValue());
+        syncB.fetchUpdates();
+        assertEquals("b", syncB.getState().getValue());
+
+        long size = syncA.bytesWrittenSinceCompaction();
+        assertTrue(size > 0);
+
+        syncA.updateState((state, updates) -> {
+            callCount.incrementAndGet();
+            updates.add(new RegularUpdate("c"));
+        });
+        assertEquals(2, callCount.get());
+        assertEquals("c", syncA.getState().getValue());
+        syncB.fetchUpdates();
+        assertEquals("c", syncB.getState().getValue());
+        assertTrue(syncA.bytesWrittenSinceCompaction() > size);
+
+        syncA.updateState((state, updates) -> {
+            callCount.incrementAndGet();
+            updates.add(new RegularUpdate("d"));
+        });
+        assertEquals(3, callCount.get());
+        assertEquals("d", syncA.getState().getValue());
+        assertEquals("c", syncB.getState().getValue());
+        syncB.fetchUpdates();
+        assertEquals("d", syncB.getState().getValue());
+        assertTrue(syncA.bytesWrittenSinceCompaction() > size);
+
+        syncA.updateState((state, updates) -> {
+            callCount.incrementAndGet();
+            updates.add(new RegularUpdate("e"));
+        });
+        assertEquals(4, callCount.get());
+        assertEquals("e", syncA.getState().getValue());
+        syncB.fetchUpdates();
+        assertEquals("e", syncB.getState().getValue());
+        assertTrue(syncA.bytesWrittenSinceCompaction() > size);
+
+        syncA.updateState((state, updates) -> {
+            callCount.incrementAndGet();
+            updates.add(new RegularUpdate("f"));
+        });
+        assertEquals(5, callCount.get());
+        assertEquals("f", syncA.getState().getValue());
+        assertTrue(syncA.bytesWrittenSinceCompaction() > size);
+
+        assertEquals("e", syncB.getState().getValue());
+        syncB.compact(state -> {
+            callCount.incrementAndGet();
+            return new RegularUpdate("e");
```

Comment:
This seems wrong. It's writing "e" to compact a state that is currently "f". Instead do `new RegularUpdate(state.getValue())`.